q2_composition/_ancombc.py

```diff
@@ -0,0 +1,159 @@
+# ----------------------------------------------------------------------------
+import subprocess
+import tempfile
+import pandas as pd
+import os
+import formulaic
+
+import qiime2
+
+from q2_stats._format import DataLoafPackageDirFmt
+
+
+def run_commands(cmds, verbose=True):
+    if verbose:
+        print('Running external command line application(s). This may print'
+              ' messages to stdout and/or stderr.')
+        print('The command(s) being run are below. These commands cannot be'
+              ' manually re-run as they will depend on temporary files that no'
+              ' longer exist.')
+    for cmd in cmds:
+        if verbose:
+            print('\nCommand:', end=' ')
+            print(' '.join(cmd), end='\n\n')
+        subprocess.run(cmd, check=True)
+
+
+def ancombc(table: pd.DataFrame, metadata: qiime2.Metadata, formula: str,
+            p_adj_method: str = 'holm', prv_cut: float = 0.1, lib_cut: int = 0,
+            level_ordering: str = None, neg_lb: bool = False,
+            tol: float = 1e-05, max_iter: int = 100, conserve: bool = False,
+            alpha: float = 0.05) -> DataLoafPackageDirFmt:
+
+    return _ancombc(
+        table=table,
+        metadata=metadata,
+        formula=formula,
+        p_adj_method=p_adj_method,
+        prv_cut=prv_cut,
+        lib_cut=lib_cut,
+        level_ordering=level_ordering,
+        neg_lb=neg_lb,
+        tol=tol,
+        max_iter=max_iter,
+        conserve=conserve,
+        alpha=alpha,
+    )
+
+
+# utility functions for formula parsing and column validation
+def _parse_terms(formula):
+    parse = formulaic.parser.parser.DefaultFormulaParser(
+        include_intercept=False)
+    terms = parse.get_ast(formula=formula).flatten()
+    formula_terms = _leaf_collector(terms)
+    return formula_terms
+
+
+def _leaf_collector(term):
+    if isinstance(term, formulaic.parser.types.Token):
+        return [str(term)]
+
+    if type(term) is not list:
+        return []
+
+    return _leaf_collector(term[1]) + _leaf_collector(term[2])
+
+
+def _column_validation(value, parameter, metadata):
+    if value not in metadata.columns:
+        raise ValueError('Value provided in the `%s` parameter was not found'
+                         ' in any of the metadata columns. Please make sure to'
+                         ' only include values that are present within the'
+                         ' metadata columns.'
+                         ' \n\n'
+                         ' Value that was not found as a metadata column: "%s"'
+                         % (parameter, value))
+
+
+def _ancombc(table, metadata, formula, p_adj_method, prv_cut, lib_cut,
+             level_ordering, neg_lb, tol, max_iter, conserve, alpha):
+
+    meta = metadata.to_dataframe()
+
+    # error on IDs found in table but not in metadata
+    missing_ids = table.index.difference(meta.index).values
+
+    if not (set(table.index).issubset(set(meta.index))):
```

Comment:
Couldn't this be `if missing_ids:`?

Reply:
Oh yeah, nice catch @Oddant1 - you're totally right! That's much easier to read 😅

Reply:
Changing this slightly to the following:
```
if missing_ids.size > 0:
```
Due to this depreciation warning (just trying to stay ahead of weird errors that will bite us in the future):
```
DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
    if missing_ids:
```